playerctl/playerctl-cli.c

```diff
@@ -16,279 +16,359 @@
+      return commands[i].func(command[1], player, error);
+    }
   }
+  g_set_error(error, playerctl_cli_error_quark (), 1, "Command not recognized: %s", command[0], NULL);
+  return FALSE;
+}
 
-  GVariant *reply = g_dbus_proxy_call_sync(proxy,
-      "ListNames",
-      NULL,
-      G_DBUS_CALL_FLAGS_NONE,
-      -1,
-      NULL,
-      &tmp_error);
+static const GOptionEntry entries[] = {
+  { "player", 'p', G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING, &player_name,
+    "The name of the player to control (default: the first available player)", "NAME" },
+  { "list-all", 'l', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, list_all_players_and_exit,
+    "List the names of running players that can be controlled and exit", NULL },
+  { "version", 'v', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, print_version_and_exit,
+    "Print version information and exit", NULL },
+  { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_STRING_ARRAY, &command, NULL, "COMMAND" },
+  { NULL }
+};
 
-  if (tmp_error != NULL) {
-    g_propagate_error(err, tmp_error);
-    g_object_unref(proxy);
-    return NULL;
+static gboolean parse_setup_options (int argc, char *argv[], GError **error)
+{
+  static const gchar *description = "Available Commands:"
+    "\n  play                    Command the player to play"
+    "\n  pause                   Command the player to pause"
+    "\n  play-pause              Command the player to toggle between play/pause"
+    "\n  stop                    Command the player to stop"
+    "\n  next                    Command the player to skip to the next track"
+    "\n  previous                Command the player to skip to the previous track"
+    "\n  position [OFFSET][+/-]  Command the player to go to the position or seek forward/backward OFFSET in seconds"
+    "\n  volume [LEVEL][+/-]     Print or set the volume to LEVEL from 0.0 to 1.0"
+    "\n  status                  Get the play status of the player"
+    "\n  metadata [KEY]          Print metadata information for the current track. If KEY is passed,"
+    "\n                          print only that value. KEY may be one of artist, title or album";
+  static const gchar *summary = "  Only for players supporting the MPRIS D-Bus specification";
+  GOptionContext *context = NULL;
+  gboolean success;
+
+  context = g_option_context_new("- Controller for MPRIS players");
+  g_option_context_add_main_entries(context, entries, NULL);
+  g_option_context_set_description(context, description);
+  g_option_context_set_summary(context, summary);
+
+  success = g_option_context_parse(context, &argc, &argv, error);
+  g_option_context_free(context);
+
+  if (!success) {
+    return FALSE;
   }
 
-  GString *names_str = g_string_new("");
-  GVariant *reply_child = g_variant_get_child_value(reply, 0);
-  gsize reply_count;
-  const gchar** names = g_variant_get_strv(reply_child, &reply_count);
+  if (command == NULL) {
+    g_set_error (error, playerctl_cli_error_quark(), 1, "No command entered");
```

Comment:
> There are no standards that I know of for return codes. Conventionally, 0 means success and >0 means error.

Pedantic, but every non-zero value means error, not just >0.[¹](http://www.faqs.org/docs/abs/HTML/exit-status.html)

> Here we interpret a command with no arguments as a command to show the help string. The help string was successfully shown, so we exit 0.

> Another interpretation is that without a command, the user told us to do nothing. In this case, we have successfully done nothing (and then some!) so we exit zero.

I see I misunderstood your intentions here. In the help string `playerctl [OPTION...] COMMAND - Controller for MPRIS players` (bear with me that I have a compiled version with the changed project description :pensive:), `COMMAND` means that a command is mandatory. If it were `[COMMAND]`, it would be optional. In such a case, it would be an error when the user does not provide a command, hence my choosing for a non-zero exit status.

Let me know if you still want me to change this or not.